piq/perceptual.py

```diff
@@ -0,0 +1,304 @@
+        self.reduction = reduction
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Content loss between feature representations of prediction and target tensors.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        """
+        _validate_input(input_tensors=(prediction, target), allow_5d=False)
+        prediction, target = _adjust_dimensions(input_tensors=(prediction, target))
+        return self.compute_metric(prediction, target)
+    
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        # Normalize input
+        mean, std = self.mean.to(prediction), self.std.to(prediction)
+        prediction = (prediction - mean) / std
+        target = (target - mean) / std
+
+        self.model.to(prediction)
+        prediction_features = self.get_features(prediction)
+        target_features = self.get_features(target)
+        
+        distances = [self.distance(x, y) for x, y in zip(prediction_features, target_features)]
+
+        # Scale distances, then average in spatial dimensions, then stack and sum in channels dimension
+        loss = torch.cat([(d * w.to(d)).mean(dim=[2, 3]) for d, w in zip(distances, self.weights)], dim=1).sum(dim=1)
+
+        if self.reduction == 'none':
+            return loss
+
+        return {'mean': loss.mean,
+                'sum': loss.sum
+                }[self.reduction](dim=0)
+
+    def get_features(self, x: torch.Tensor) -> List[torch.Tensor]:
+        """
+        Args:
+            x: torch.Tensor with shape (N, C, H, W)
+        
+        Returns:
+            features: List of features extracted from intermediate layers
+        """
+        features = []
+        for name, module in self.model._modules.items():
+            x = module(x)
+            if name in self.layers:
+                features.append(self.normalize(x) if self.normalize_features else x)
+        return features
+
+    def normalize(self, x: torch.Tensor) -> torch.Tensor:
+        """Normalize feature maps in channel direction to unit length.
+        Args:
+            x: Tensor with shape (N, C, H, W)
+        Returns:
+            x_norm: Normalized input
+        """
+        norm_factor = torch.sqrt(torch.sum(x ** 2, dim=1, keepdim=True))
+        return x / (norm_factor + self.EPS)
+
+    def max_pool_to_average_pool(self, module: torch.nn.Module) -> torch.nn.Module:
+        r"""Turn All MaxPool layers into AveragePool"""
+        module_output = module
+        if isinstance(module, torch.nn.MaxPool2d):
+            module_output = torch.nn.AvgPool2d(kernel_size=2, stride=2, padding=0)
+            
+        for name, child in module.named_children():
+            module_output.add_module(name, self.max_pool_to_average_pool(child))
+        del module
+        return module_output
```

Comment:
I don't understand why you first copy the reference to the `module` object, then modify it and then delete the initial reference in case if you reassign the output to the same object (`self.model = self.max_pool_to_average_pool(self.model)`, line 130) anyway.
To me it seems like deleting lines 211 and 217 changes nothing. Can you comment on that?

Reply:
You are right, `del` isn't necessary. 
Without copy it fails when tries to call `add_module` for not existing `module_output` variable.